Test/Unit/Block/Catalog/Product/ViewedProductTest.php

```diff
@@ -0,0 +1,223 @@
+        $productMock->method('getPriceInfo')->willReturn($priceInfoMock);
+
+        $categoryMock0 = $this->createMock(Category::class);
+        $categoryMock0->method('getName')->willReturn(self::PRODUCT_CATEGORY_NAMES[0]);
+        $this->categoryMocks[0] = $categoryMock0;
+
+        $categoryMock1 = $this->createMock(Category::class);
+        $categoryMock1->method('getName')->willReturn(self::PRODUCT_CATEGORY_NAMES[1]);
+        $this->categoryMocks[1] = $categoryMock1;
+
+        $categoryMock2 = $this->createMock(Category::class);
+        $categoryMock2->method('getName')->willReturn(self::PRODUCT_CATEGORY_NAMES[2]);
+        $this->categoryMocks[2] = $categoryMock2;
+
+        $categoryMock3 = $this->createMock(Category::class);
+        $categoryMock3->method('getName')->willReturn(self::PRODUCT_CATEGORY_NAMES[3]);
+        $this->categoryMocks[3] = $categoryMock3;
+
+        $categoryMock4 = $this->createMock(Category::class);
+        $categoryMock4->method('getName')->willReturn(self::PRODUCT_CATEGORY_NAMES[4]);
+        $this->categoryMocks[4] = $categoryMock4;
+
+        $categoryCollectionMock = $this->createMock(CategoryCollection::class);
+        $categoryCollectionMock->method('load')->will($this->returnCallback(
+            function ($id) {
+                switch ($id) {
+                    case self::PRODUCT_CATEGORY_IDS[0];
+                        return $this->categoryMocks[0];
+                        break;
+                    case self::PRODUCT_CATEGORY_IDS[1];
+                        return $this->categoryMocks[1];
+                        break;
+                    case self::PRODUCT_CATEGORY_IDS[2];
+                        return $this->categoryMocks[2];
+                        break;
+                    case self::PRODUCT_CATEGORY_IDS[3];
+                        return $this->categoryMocks[3];
+                        break;
+                    case self::PRODUCT_CATEGORY_IDS[4];
+                        return $this->categoryMocks[4];
+                        break;
+                }
+            }
+        ));
+        $categoryFactoryMock = $this->getMockBuilder(CategoryFactory::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['create'])
+            ->getMock();
+        $categoryFactoryMock->method('create')->willReturn($categoryCollectionMock);
+
+        $productMock->method('getCategoryIds')->willReturn(self::PRODUCT_CATEGORY_IDS);
+
+        $registryMock = $this->createMock(Registry::class);
+        $registryMock->method('registry')->willReturn($productMock);
+
+        $imageMock = $this->createMock(Image::class);
+        $imageMock->method('init')
+            ->with($this->isInstanceOf(Product::class),$this->equalTo('product_base_image'))
+            ->willReturn($imageMock);
+        $imageMock->method('getUrl')->willReturn(self::PRODUCT_IMAGE_URL);
+
+        $this->object = new ViewedProduct(
```

Comment:
is there a better name than object for this? it's kinda weird imo to call it object when it's a viewed product?